crates/bevy_ecs/src/bundle.rs

```diff
@@ -571,192 +501,379 @@ impl BundleInfo {
+                        } else {
+                            // SAFETY: the only two borrowed archetypes are above and we just did collision checks
+                            &mut *archetypes_ptr.add(swapped_location.archetype_id.index())
+                        };
 
-                    self.entities.set(
-                        swapped_entity.index(),
-                        EntityLocation {
-                            archetype_id: swapped_location.archetype_id,
-                            archetype_row: swapped_location.archetype_row,
-                            table_id: swapped_location.table_id,
-                            table_row: result.table_row,
-                        },
+                        entities.set(
+                            swapped_entity.index(),
+                            EntityLocation {
+                                archetype_id: swapped_location.archetype_id,
+                                archetype_row: swapped_location.archetype_row,
+                                table_id: swapped_location.table_id,
+                                table_row: result.table_row,
+                            },
+                        );
+                        swapped_archetype
+                            .set_entity_table_row(swapped_location.archetype_row, result.table_row);
+                    }
+
+                    let bundle_info = &*self.bundle_info;
+                    let add_bundle = &*self.add_bundle;
+                    bundle_info.write_components(
+                        new_table,
+                        sparse_sets,
+                        add_bundle,
+                        entity,
+                        move_result.new_row,
+                        self.change_tick,
+                        bundle,
                     );
-                    swapped_archetype
-                        .set_entity_table_row(swapped_location.archetype_row, result.table_row);
-                }
 
-                // PERF: this could be looked up during Inserter construction and stored (but borrowing makes this nasty)
-                // SAFETY: The edge is assured to be initialized when creating the BundleInserter
-                let add_bundle = unsafe {
-                    self.archetype
-                        .edges()
-                        .get_add_bundle_internal(self.bundle_info.id)
-                        .debug_checked_unwrap()
+                    new_location
                 };
-                self.bundle_info.write_components(
-                    new_table,
-                    self.sparse_sets,
-                    add_bundle,
-                    entity,
-                    move_result.new_row,
-                    self.change_tick,
-                    bundle,
-                );
+
+                // SAFETY: We have no oustanding mutable references to world as they were dropped
+                trigger_hooks(&**new_archetype, self.world.into_deferred());
```

Comment:
This has a safety comment but no `unsafe`  block.

Reply:
There are many instances of that in the surrounding code, both de-referencing the pointer and `into_deferred` are unsafe operations so it's there to explain those. We can also add redundant unsafe blocks which also exist nearby. This is mostly a style decision so I'm happy with whatever is preferred.

Reply:
Given the recent direction to annotate all of our `unsafe` calls with `unsafe`, even in `unsafe` functions, I think adding them would be for the best. It'd help out with #11590 too.